app/src/main/java/ch/hikemate/app/model/route/HikesViewModel.kt

```diff
@@ -0,0 +1,1166 @@
+        var successful = false
+        _hikesMutex.withLock {
+          // Retrieve the hike to select from the loaded hikes
+          val hike = _hikeFlowsMap[hikeId] ?: return@withLock
+
+          // Mark the hike as selected
+          _selectedHikeId = hikeId
+          _selectedHike.value = hike.value
+          successful = true
+        }
+
+        // Perform the callback without the lock, to avoid deadlocks and improve performance
+        if (successful) {
+          onSuccess()
+        } else {
+          onFailure()
+        }
+      }
+
+  private suspend fun unselectHikeAsync() =
+      _hikesMutex.withLock {
+        // Only emit null as a value if the selected hike was not already null
+        if (_selectedHikeId != null) {
+          _selectedHikeId = null
+          _selectedHike.value = null
+        }
+      }
+
+  /**
+   * This function performs a loading operation but DOES NOT set [_loading]. If calling it from
+   * another function, please set [_loading] to true before calling the function and to false once
+   * the call has ended.
+   *
+   * @return True if the operation is successful, false otherwise.
+   */
+  private suspend fun refreshSavedHikesCacheAsync(
+      onSuccess: () -> Unit = {},
+      onFailure: () -> Unit = {}
+  ): Boolean =
+      withContext(dispatcher) {
+        val savedHikes: List<SavedHike>
+        try {
+          // Load the saved hikes from the repository
+          savedHikes = savedHikesRepo.loadSavedHikes()
+        } catch (e: Exception) {
+          Log.e(LOG_TAG, "Error encountered while loading saved hikes", e)
+          onFailure()
+          return@withContext false
+        }
+
+        // Wait for the lock to avoid concurrent modifications
+        _hikesMutex.withLock { updateSavedHikesCache(savedHikes) }
+
+        onSuccess()
+        return@withContext true
+      }
+
+  /**
+   * Helper function for [refreshSavedHikesCacheAsync].
+   *
+   * Takes a list of saved hikes and updates the local cache of saved hikes with that list.
+   *
+   * It is the caller's responsibility to ensure the saved hikes list is up-to-date and to acquire
+   * [_hikesMutex] before calling this function.
+   *
+   * @param newList The list of saved hikes to update the cache with.
+   */
+  private fun updateSavedHikesCache(newList: List<SavedHike>) {
+    // Clear the current saved hikes register
+    _savedHikesMap.clear()
+
+    // Build the new saved hikes register
+    newList.forEach { savedHike -> _savedHikesMap[savedHike.id] = savedHike }
+
+    // Update the map of hikes to reflect the changes
+    val toRemove: MutableSet<String> = mutableSetOf()
+    _hikeFlowsMap.values.forEach { hikeFlow ->
+      val hike = hikeFlow.value
+      val savedHike = _savedHikesMap[hike.id]
+      // If the list currently only contains saved hikes and this hike was unsaved, remove it
+      if (_loadedHikesType == LoadedHikes.FromSaved && savedHike == null) {
+        toRemove.add(hike.id)
```

Comment:
Yes it's because of the `forEach`. More specifically, I had bad experiences in the past when changing a collection while iterating on it.

I could use an iterator, which would allow me to have more granular control, but I believe it would make the code less readable.

However I'll give that to you, it is not very clean. Maybe I should try with the iterator anyway?

Reply:
To be clear : I did not test using `remove` in the `forEach`, but from my experience and from what I've seen online, it's not good practice and is very likely to fail.

Reply:
Yep, I've got bad time with it before too. A set of string is good to me, for performance, reliability and any other worries we have!